{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "pleasant-birthday",
   "metadata": {},
   "source": [
    "# Amplicon mapping to AGORA2 Genomes\n",
    "## Notebook 2: Creating the ASV to genome mapping file\n",
    "\n",
    "#### Kathleen Beilsmith, Argonne National Lab (ANL)\n",
    "#### ANL Henry Group & UChicago Chang Lab Mouse Gut Microbiome collaboration\n",
    "#### Data Credit: Megan Kennedy"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "separated-juvenile",
   "metadata": {},
   "source": [
    "This notebook takes a compendium of marker gene sequences found in a collection of genomes (from Notebook 1) and compares them to representative sequences from an amplicon dataset.\n",
    "\n",
    "Inputs:\n",
    "- JSON file with marker gene compendium\n",
    "- fasta file with amplicon sequence variant (ASV) sequences (from dada2 or other denoising pipeline)\n",
    "- forward and reverse primer sequences used to generate amplicons\n",
    "\n",
    "**ePCR()** uses the marker gene compendium and primer sequences to perform in silico PCR and cut out the region of interest for each copy of the marker gene (eAmplicon).\n",
    "\n",
    "**fuzzy_match_amplicons_to_genomes()** compares each ASV from the dataset to the set of eAmplicons generated with the primers and returns matches above a specified sequence identity threshold. The result is a JSON mapping file with one entry per ASV with a list of names for genomes containing matching eAmplicons.\n",
    "\n",
    "Sequence alignment scores are generated with Biopython Align.PairwiseAligner() with 'global' mode. The match and mismatch scores are 1 and 0, respectively, and the penalties for opening and extending gaps are -1 and -0.5, respectively. \n",
    "https://biopython.org/docs/1.75/api/Bio.Align.html#Bio.Align.PairwiseAlignment\n",
    "With these settings, the rounded **alignment score** corresponds to the number of nucleotides matching in the sequences. \n",
    "\n",
    "The **expected** number of nucleotide matches at the given identity threshold is calculated as the identity threshold (float) x amplicon sequence length (int), rounded.\n",
    "\n",
    "If the alignment score is greater than or equal to the expected score, the amplicon from the dataset is considered matched to an eAmplicon from the compendium (AGORA2) at the identity threshold.\n",
    "\n",
    "**best_match_amplicons_to_genomes()** uses fuzzy_match_amplicons_to_genomes() to get results at every identity threshold between 1 and a provided lower bound. The result is a JSON mapping file with one entry per ASV with a list of names for genomes containing matching eAmplicons. Only the results for the highest similarity level at which a match was found are stored."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "crazy-balance",
   "metadata": {},
   "outputs": [],
   "source": [
    "############################\n",
    "# Required modules\n",
    "############################\n",
    "\n",
    "import numpy as np\n",
    "\n",
    "import os\n",
    "import modelseedpy\n",
    "import json as _json\n",
    "from modelseedpy import RPCClient\n",
    "from Bio import SeqUtils\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "\n",
    "import json\n",
    "import glob\n",
    "\n",
    "from Bio.Align import MultipleSeqAlignment\n",
    "from Bio import Align\n",
    "from Bio import pairwise2 "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "wrapped-strike",
   "metadata": {},
   "source": [
    "### Use compendium of marker gene copies and primer sequences to do ePCR."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "cellular-placement",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('/Users/kbeilsmith/Desktop/2024_AmtG/AmtG_0.0.2_test_files/test_Compendium.json') as json_file:\n",
    "    rRNA_dict = json.load(json_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "twelve-carter",
   "metadata": {},
   "outputs": [],
   "source": [
    "############################\n",
    "# Primer sequences\n",
    "############################\n",
    "\n",
    "f_pr = Seq('AYTGGGYDTAAAGNG')\n",
    "r_pr = Seq('CCGTCAATTYHTTTRAGT')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "shared-trance",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Given a marker gene compendium and primers\n",
    "# this function returns an eAmplicon sequence for each marker gene copy,\n",
    "# with the information about the location (genome) of the marker gene copy.\n",
    "\n",
    "def ePCR(f_primer,r_primer,marker_gene_compendium):\n",
    "    \n",
    "    eAmplicon_set = {}\n",
    "    \n",
    "    # For each marker gene copy in the compendium:\n",
    "    for feature in marker_gene_compendium:\n",
    "    # print(rRNA_dict[feature])\n",
    "    # print(feature)\n",
    "        \n",
    "        # Find exact sequence matches with the F and then R primers.\n",
    "        fwd_ePCR = SeqUtils.nt_search(str(marker_gene_compendium[feature]),f_primer)\n",
    "        rev_ePCR = SeqUtils.nt_search(str(marker_gene_compendium[feature]),r_primer.reverse_complement())\n",
    "        \n",
    "        # If both primers match:\n",
    "        if len(fwd_ePCR) >= 2 and len(rev_ePCR) >=2:\n",
    "            print(f\"Marker gene copy: {feature}\")\n",
    "            print(f\"Forward primer match: {fwd_ePCR}\")\n",
    "            print(f\"Reverse primer match: {rev_ePCR}\")\n",
    "            # Add primer length to find start of amplicon sequence from forward match.\n",
    "            start_cut = fwd_ePCR[1]+len(f_primer)\n",
    "            # The beginning of the reverse primer match is the end of the amplicon.\n",
    "            end_cut = rev_ePCR[1]\n",
    "            # Use the start and end positions to get the eAmplicon sequence.\n",
    "            full_feature_seq = marker_gene_compendium[feature]\n",
    "            eAmplicon = full_feature_seq[start_cut:end_cut]\n",
    "            eAmplicon_length = len(eAmplicon)\n",
    "            print(f\"eAmplicon length: {eAmplicon_length}\")\n",
    "            print(f\"eAmplicon sequence: {eAmplicon} \\n\")\n",
    "        \n",
    "        # Store the eAmplicon with the marker gene record of genome, contig, direction, etc.\n",
    "        eAmplicon_set[feature] = eAmplicon\n",
    "        \n",
    "    return eAmplicon_set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "specialized-democrat",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Marker gene copy: GCA_000210035.1_ASM21003v1_genomic.fna;FP929055.1;+;147;1539\n",
      "Forward primer match: ['A[CT]TGGG[CT][AGT]TAAAG[GATC]G', 555]\n",
      "Reverse primer match: ['ACT[CT]AAA[AGT][AG]AATTGACGG', 901]\n",
      "eAmplicon length: 331\n",
      "eAmplicon sequence: AGCGTAGACGGAGTGGCAAGTCTGATGTGAAAACCCGGGGCTCAACCCCGGGACTGCATTGGAAACTGTCAATCTGGAGTACCGGAGAGGTAAGCGGAATTCCTAGTGTAGCGGTGAAATGCGTAGATATTAGGAGGAACACCAGTGGCGAAGGCGGCTTACTGGACGGTAACTGACGTTGAGGCTCGAAAGCGTGGGGAGCAAACAGGATTAGATACCCTGGTAGTCCACGCCGTAAACGATGACTACTAGGTGTCGGGCAGCAAAGCTGTTCGGTGCCGCAGCAAACGCAATAAGTAGTCCACCTGGGGAGTACGTTCGCAAGAATGAA \n",
      "\n",
      "Marker gene copy: GCA_000025225.2_ASM2522v1_genomic.fna;CP001850.2;+;255524;1534\n",
      "Forward primer match: ['A[CT]TGGG[CT][AGT]TAAAG[GATC]G', 551]\n",
      "Reverse primer match: ['ACT[CT]AAA[AGT][AG]AATTGACGG', 898]\n",
      "eAmplicon length: 332\n",
      "eAmplicon sequence: CGTGCAGGCGGGCTGATAAGTCAGATGTGAAATCCCCGAGCTTAACTCGGGAACTGCATCTGATACTGTTGGTCTTGAGTGCTGGAGAGGATAGTGGAATTCCTAGTGTAGCGGTAAAATGCGCAGATATTAGGAGGAACACCAGTGGCGAAGGCGGCTATCTGGACAGTAACTGACGCTGAGGCGCGAAAGCGTGGGTAGCAAACAGGATTAGATACCCTGGTAGTCCACGCCGTAAACGATGATTACTAGGTGTAGGAGGTATCGACCCCTTCTGTGCCGGAGTTAACACAATAAGTAATCCACCTGGGGAGTACGGCCGCAAGGTTGAA \n",
      "\n",
      "Marker gene copy: GCA_000025225.2_ASM2522v1_genomic.fna;CP001850.2;+;465371;1534\n",
      "Forward primer match: ['A[CT]TGGG[CT][AGT]TAAAG[GATC]G', 551]\n",
      "Reverse primer match: ['ACT[CT]AAA[AGT][AG]AATTGACGG', 898]\n",
      "eAmplicon length: 332\n",
      "eAmplicon sequence: CGTGCAGGCGGGCTGATAAGTCAGATGTGAAATCCCCGAGCTTAACTCGGGAACTGCATCTGATACTGTTGGTCTTGAGTGCTGGAGAGGATAGTGGAATTCCTAGTGTAGCGGTAAAATGCGCAGATATTAGGAGGAACACCAGTGGCGAAGGCGGCTATCTGGACAGTAACTGACGCTGAGGCGCGAAAGCGTGGGTAGCAAACAGGATTAGATACCCTGGTAGTCCACGCCGTAAACGATGATTACTAGGTGTAGGAGGTATCGACCCCTTCTGTGCCGGAGTTAACACAATAAGTAATCCACCTGGGGAGTACGGCCGCAAGGTTGAA \n",
      "\n",
      "Marker gene copy: GCA_000189435.3_ASM18943v3_genomic.fna;CP024649.1;+;531515;1558\n",
      "Forward primer match: ['A[CT]TGGG[CT][AGT]TAAAG[GATC]G', 571]\n",
      "Reverse primer match: ['ACT[CT]AAA[AGT][AG]AATTGACGG', 918]\n",
      "eAmplicon length: 332\n",
      "eAmplicon sequence: CGCGTAGGCGGTTTTTTAAGTCTGATGTGAAAGCCCACGGCTCAACCGTGGAGGGTCATTGGAAACTGGAAAACTTGAGTGCAGAAGAGGAAAGTGGAATTCCATGTGTAGCGGTGAAATGCGCAGAGATATGGAGGAACACCAGTGGCGAAGGCGACTTTCTGGTCTGTAACTGACGCTGATGTGCGAAAGCGTGGGGATCAAACAGGATTAGATACCCTGGTAGTCCACGCCGTAAACGATGAGTGCTAAGTGTTAGGGGGTTTCCGCCCCTTAGTGCTGCAGCTAACGCATTAAGCACTCCGCCTGGGGAGTACGACCGCAAGGTTGAA \n",
      "\n",
      "Marker gene copy: GCA_000189435.3_ASM18943v3_genomic.fna;CP024649.1;-;1908241;1565\n",
      "Forward primer match: ['A[CT]TGGG[CT][AGT]TAAAG[GATC]G', 579]\n",
      "Reverse primer match: ['ACT[CT]AAA[AGT][AG]AATTGACGG', 926]\n",
      "eAmplicon length: 332\n",
      "eAmplicon sequence: CGCGTAGGCGGTTTTTTAAGTCTGATGTGAAAGCCCACGGCTCAACCGTGGAGGGTCATTGGAAACTGGAAAACTTGAGTGCAGAAGAGGAAAGTGGAATTCCATGTGTAGCGGTGAAATGCGCAGAGATATGGAGGAACACCAGTGGCGAAGGCGACTTTCTGGTCTGTAACTGACGCTGATGTGCGAAAGCGTGGGGATCAAACAGGATTAGATACCCTGGTAGTCCACGCCGTAAACGATGAGTGCTAAGTGTTAGGGGGTTTCCGCCCCTTAGTGCTGCAGCTAACGCATTAAGCACTCCGCCTGGGGAGTACGACCGCAAGGTTGAA \n",
      "\n",
      "Marker gene copy: GCA_000189435.3_ASM18943v3_genomic.fna;CP024649.1;-;2090043;1565\n",
      "Forward primer match: ['A[CT]TGGG[CT][AGT]TAAAG[GATC]G', 579]\n",
      "Reverse primer match: ['ACT[CT]AAA[AGT][AG]AATTGACGG', 926]\n",
      "eAmplicon length: 332\n",
      "eAmplicon sequence: CGCGTAGGCGGTTTTTTAAGTCTGATGTGAAAGCCCACGGCTCAACCGTGGAGGGTCATTGGAAACTGGAAAACTTGAGTGCAGAAGAGGAAAGTGGAATTCCATGTGTAGCGGTGAAATGCGCAGAGATATGGAGGAACACCAGTGGCGAAGGCGACTTTCTGGTCTGTAACTGACGCTGATGTGCGAAAGCGTGGGGATCAAACAGGATTAGATACCCTGGTAGTCCACGCCGTAAACGATGAGTGCTAAGTGTTAGGGGGTTTCCGCCCCTTAGTGCTGCAGCTAACGCATTAAGCACTCCGCCTGGGGAGTACGACCGCAAGGTTGAA \n",
      "\n",
      "Marker gene copy: GCA_000189435.3_ASM18943v3_genomic.fna;CP024649.1;-;2205931;1564\n",
      "Forward primer match: ['A[CT]TGGG[CT][AGT]TAAAG[GATC]G', 579]\n",
      "Reverse primer match: ['ACT[CT]AAA[AGT][AG]AATTGACGG', 926]\n",
      "eAmplicon length: 332\n",
      "eAmplicon sequence: CGCGTAGGCGGTTTTTTAAGTCTGATGTGAAAGCCCACGGCTCAACCGTGGAGGGTCATTGGAAACTGGAAAACTTGAGTGCAGAAGAGGAAAGTGGAATTCCATGTGTAGCGGTGAAATGCGCAGAGATATGGAGGAACACCAGTGGCGAAGGCGACTTTCTGGTCTGTAACTGACGCTGATGTGCGAAAGCGTGGGGATCAAACAGGATTAGATACCCTGGTAGTCCACGCCGTAAACGATGAGTGCTAAGTGTTAGGGGGTTTCCGCCCCTTAGTGCTGCAGCTAACGCATTAAGCACTCCGCCTGGGGAGTACGACCGCAAGGTTGAA \n",
      "\n",
      "Marker gene copy: GCF_000157535.1_ASM15753v1_genomic.fna;NZ_GG688487.1;-;2137;1573\n",
      "Forward primer match: ['A[CT]TGGG[CT][AGT]TAAAG[GATC]G', 590]\n",
      "Reverse primer match: ['ACT[CT]AAA[AGT][AG]AATTGACGG', 937]\n",
      "eAmplicon length: 332\n",
      "eAmplicon sequence: AGCGCAGGCGGTTTCTTAAGTCTGATGTGAAAGCCCCCGGCTCAACCGGGGAGGGTCATTGGAAACTGGGAGACTTGAGTGCAGAAGAGGAGAGTGGAATTCCATGTGTAGCGGTGAAATGCGTAGATATATGGAGGAACACCAGTGGCGAAGGCGGCTCTCTGGTCTGTAACTGACGCTGAGGCTCGAAAGCGTGGGGAGCAAACAGGATTAGATACCCTGGTAGTCCACGCCGTAAACGATGAGTGCTAAGTGTTGGAGGGTTTCCGCCCTTCAGTGCTGCAGCTAACGCATTAAGCACTCCGCCTGGGGAGTACGACCGCAAGGTTGAA \n",
      "\n",
      "Marker gene copy: GCF_000157555.1_ASM15755v1_genomic.fna;NZ_GG688445.1;+;133;1573\n",
      "Forward primer match: ['A[CT]TGGG[CT][AGT]TAAAG[GATC]G', 589]\n",
      "Reverse primer match: ['ACT[CT]AAA[AGT][AG]AATTGACGG', 936]\n",
      "eAmplicon length: 332\n",
      "eAmplicon sequence: AGCGCAGGCGGTTTCTTAAGTCTGATGTGAAAGCCCCCGGCTCAACCGGGGAGGGTCATTGGAAACTGGGAGACTTGAGTGCAGAAGAGGAGAGTGGAATTCCATGTGTAGCGGTGAAATGCGTAGATATATGGAGGAACACCAGTGGCGAAGGCGGCTCTCTGGTCTGTAACTGACGCTGAGGCTCGAAAGCGTGGGGAGCAAACAGGATTAGATACCCTGGTAGTCCACGCCGTAAACGATGAGTGCTAAGTGTTGGAGGGTTTCCGCCCTTCAGTGCTGCAGCTAACGCATTAAGCACTCCGCCTGGGGAGTACGACCGCAAGGTTGAA \n",
      "\n",
      "Marker gene copy: GCF_001689125.2_ASM168912v2_genomic.fna;NZ_CP015405.2;+;196245;1539\n",
      "Forward primer match: ['A[CT]TGGG[CT][AGT]TAAAG[GATC]G', 553]\n",
      "Reverse primer match: ['ACT[CT]AAA[AGT][AG]AATTGACGG', 899]\n",
      "eAmplicon length: 331\n",
      "eAmplicon sequence: AGCGTAGACGGAAGAGCAAGTCTGATGTGAAAGGCTGGGGCTTAACCCCAGGACTGCATTGGAAACTGTTTTTCTAGAGTGCCGGAGAGGTAAGCGGAATTCCTAGTGTAGCGGTGAAATGCGTAGATATTAGGAGGAACACCAGTGGCGAAGGCGGCTTACTGGACGGTAACTGACGTTGAGGCTCGAAAGCGTGGGGAGCAAACAGGATTAGATACCCTGGTAGTCCACGCCGTAAACGATGAATACTAGGTGTCGGGTGGCAAAGCCATTCGGTGCCGCAGCAAACGCAATAAGTATTCCACCTGGGGAGTACGTTCGCAAGAATGAA \n",
      "\n",
      "Marker gene copy: GCF_001689125.2_ASM168912v2_genomic.fna;NZ_CP015405.2;+;210897;1539\n",
      "Forward primer match: ['A[CT]TGGG[CT][AGT]TAAAG[GATC]G', 553]\n",
      "Reverse primer match: ['ACT[CT]AAA[AGT][AG]AATTGACGG', 899]\n",
      "eAmplicon length: 331\n",
      "eAmplicon sequence: AGCGTAGACGGAAGAGCAAGTCTGATGTGAAAGGCTGGGGCTTAACCCCAGGACTGCATTGGAAACTGTTTTTCTAGAGTGCCGGAGAGGTAAGCGGAATTCCTAGTGTAGCGGTGAAATGCGTAGATATTAGGAGGAACACCAGTGGCGAAGGCGGCTTACTGGACGGTAACTGACGTTGAGGCTCGAAAGCGTGGGGAGCAAACAGGATTAGATACCCTGGTAGTCCACGCCGTAAACGATGAATACTAGGTGTCGGGTGGCAAAGCCATTCGGTGCCGCAGCAAACGCAATAAGTATTCCACCTGGGGAGTACGTTCGCAAGAATGAA \n",
      "\n",
      "Marker gene copy: GCF_001689125.2_ASM168912v2_genomic.fna;NZ_CP015405.2;-;2331707;1539\n",
      "Forward primer match: ['A[CT]TGGG[CT][AGT]TAAAG[GATC]G', 554]\n",
      "Reverse primer match: ['ACT[CT]AAA[AGT][AG]AATTGACGG', 900]\n",
      "eAmplicon length: 331\n",
      "eAmplicon sequence: AGCGTAGACGGAAGAGCAAGTCTGATGTGAAAGGCTGGGGCTTAACCCCAGGACTGCATTGGAAACTGTTTTTCTAGAGTGCCGGAGAGGTAAGCGGAATTCCTAGTGTAGCGGTGAAATGCGTAGATATTAGGAGGAACACCAGTGGCGAAGGCGGCTTACTGGACGGTAACTGACGTTGAGGCTCGAAAGCGTGGGGAGCAAACAGGATTAGATACCCTGGTAGTCCACGCCGTAAACGATGAATACTAGGTGTCGGGTGGCAAAGCCATTCGGTGCCGCAGCAAACGCAATAAGTATTCCACCTGGGGAGTACGTTCGCAAGAATGAA \n",
      "\n",
      "Marker gene copy: GCF_001689125.2_ASM168912v2_genomic.fna;NZ_CP015405.2;-;3598716;1539\n",
      "Forward primer match: ['A[CT]TGGG[CT][AGT]TAAAG[GATC]G', 554]\n",
      "Reverse primer match: ['ACT[CT]AAA[AGT][AG]AATTGACGG', 900]\n",
      "eAmplicon length: 331\n",
      "eAmplicon sequence: AGCGTAGACGGAAGAGCAAGTCTGATGTGAAAGGCTGGGGCTTAACCCCAGGACTGCATTGGAAACTGTTTTTCTAGAGTGCCGGAGAGGTAAGCGGAATTCCTAGTGTAGCGGTGAAATGCGTAGATATTAGGAGGAACACCAGTGGCGAAGGCGGCTTACTGGACGGTAACTGACGTTGAGGCTCGAAAGCGTGGGGAGCAAACAGGATTAGATACCCTGGTAGTCCACGCCGTAAACGATGAATACTAGGTGTCGGGTGGCAAAGCCATTCGGTGCCGCAGCAAACGCAATAAGTATTCCACCTGGGGAGTACGTTCGCAAGAATGAA \n",
      "\n",
      "Marker gene copy: GCF_001689125.2_ASM168912v2_genomic.fna;NZ_CP015405.2;-;4239140;1539\n",
      "Forward primer match: ['A[CT]TGGG[CT][AGT]TAAAG[GATC]G', 554]\n",
      "Reverse primer match: ['ACT[CT]AAA[AGT][AG]AATTGACGG', 900]\n",
      "eAmplicon length: 331\n",
      "eAmplicon sequence: AGCGTAGACGGAAGAGCAAGTCTGATGTGAAAGGCTGGGGCTTAACCCCAGGACTGCATTGGAAACTGTTTTTCTAGAGTGCCGGAGAGGTAAGCGGAATTCCTAGTGTAGCGGTGAAATGCGTAGATATTAGGAGGAACACCAGTGGCGAAGGCGGCTTACTGGACGGTAACTGACGTTGAGGCTCGAAAGCGTGGGGAGCAAACAGGATTAGATACCCTGGTAGTCCACGCCGTAAACGATGAATACTAGGTGTCGGGTGGCAAAGCCATTCGGTGCCGCAGCAAACGCAATAAGTATTCCACCTGGGGAGTACGTTCGCAAGAATGAA \n",
      "\n",
      "Marker gene copy: GCA_000185445.1_ASM18544v1_genomic.fna;GL622347.1;+;171;1546\n",
      "Forward primer match: ['A[CT]TGGG[CT][AGT]TAAAG[GATC]G', 557]\n",
      "Reverse primer match: ['ACT[CT]AAA[AGT][AG]AATTGACGG', 906]\n",
      "eAmplicon length: 334\n",
      "eAmplicon sequence: CTCGTAGGTGGTTCGTCGCGTCTGTCGTGAAAGCCAGCAGCTTAACTGTTGGTCTGCGGTGGGTACGGGCGGGCTTGAGTGCGGTAGGGGTGACTGGAATTCCTGGTGTAGCGGTGGAATGCGCAGATATCAGGAGGAACACCGATGGCGAAGGCAGGTCACTGGGCCGTTACTGACACTGAGGAGCGAAAGCGTGGGGAGCGAACAGGATTAGATACCCTGGTAGTCCACGCTGTAAACGTTGGGAACTAGGTGTGGGGATGCTATCCTGTGTTTCTGCGCCGTAGCTAACGCATTAAGTTCCCCGCCTGGGGAGTACGGTCGCAAGGCTAAA \n",
      "\n",
      "Marker gene copy: GCA_000175395.3_ASM17539v3_genomic.fna;ACSJ01000010.1;+;197;1519\n",
      "Forward primer match: ['A[CT]TGGG[CT][AGT]TAAAG[GATC]G', 538]\n",
      "Reverse primer match: ['ACT[CT]AAA[AGT][AG]AATTGACGG', 883]\n",
      "eAmplicon length: 330\n",
      "eAmplicon sequence: TATGTAGGTGGGTGCTTAAGTCAGATGTGAAATTCCCGGGCTTAACCTGGGCGCTGCATTTGAAACTGGGCATCTAGAGTGCAGGAGAGGAAAGTGGAATTCCTAGTGTAGCGGTGAAATGCGTAGAGATTAGGAAGAACACCAGTGGCGAAGGCGACTTTCTGGACTGTAACTGACACTGAGATACGAAAGCGTGGGTAGCAAACAGGATTAGATACCCTGGTAGTCCACGCCGTAAACGATGAATACTAGGTGTCGGGGGGTACCACCCTCGGTGCCGCAGCAAACGCATTAAGTATTCCGCCTGGGGAGTACGGTCGCAAGATTAAA \n",
      "\n",
      "Marker gene copy: GCA_000175395.3_ASM17539v3_genomic.fna;ACSJ01000009.1;+;100315;1519\n",
      "Forward primer match: ['A[CT]TGGG[CT][AGT]TAAAG[GATC]G', 538]\n",
      "Reverse primer match: ['ACT[CT]AAA[AGT][AG]AATTGACGG', 883]\n",
      "eAmplicon length: 330\n",
      "eAmplicon sequence: TATGTAGGTGGGTGCTTAAGTCAGATGTGAAATTCCCGGGCTTAACCTGGGCGCTGCATTTGAAACTGGGCATCTAGAGTGCAGGAGAGGAAAGTGGAATTCCTAGTGTAGCGGTGAAATGCGTAGAGATTAGGAAGAACACCAGTGGCGAAGGCGACTTTCTGGACTGTAACTGACACTGAGATACGAAAGCGTGGGTAGCAAACAGGATTAGATACCCTGGTAGTCCACGCCGTAAACGATGAATACTAGGTGTCGGGGGGTACCACCCTCGGTGCCGCAGCAAACGCATTAAGTATTCCGCCTGGGGAGTACGGTCGCAAGATTAAA \n",
      "\n",
      "Marker gene copy: GCA_000175395.3_ASM17539v3_genomic.fna;ACSJ01000008.1;+;20339;1519\n",
      "Forward primer match: ['A[CT]TGGG[CT][AGT]TAAAG[GATC]G', 538]\n",
      "Reverse primer match: ['ACT[CT]AAA[AGT][AG]AATTGACGG', 883]\n",
      "eAmplicon length: 330\n",
      "eAmplicon sequence: TATGTAGGTGGGTGCTTAAGTCAGATGTGAAATTCCCGGGCTTAACCTGGGCGCTGCATTTGAAACTGGGCATCTAGAGTGCAGGAGAGGAAAGTGGAATTCCTAGTGTAGCGGTGAAATGCGTAGAGATTAGGAAGAACACCAGTGGCGAAGGCGACTTTCTGGACTGTAACTGACACTGAGATACGAAAGCGTGGGTAGCAAACAGGATTAGATACCCTGGTAGTCCACGCCGTAAACGATGAATACTAGGTGTCGGGGGGTACCACCCTCGGTGCCGCAGCAAACGCATTAAGTATTCCGCCTGGGGAGTACGGTCGCAAGATTAAA \n",
      "\n",
      "Marker gene copy: GCA_000175395.3_ASM17539v3_genomic.fna;ACSJ01000003.1;+;8897;1519\n",
      "Forward primer match: ['A[CT]TGGG[CT][AGT]TAAAG[GATC]G', 538]\n",
      "Reverse primer match: ['ACT[CT]AAA[AGT][AG]AATTGACGG', 883]\n",
      "eAmplicon length: 330\n",
      "eAmplicon sequence: TATGTAGGTGGGTGCTTAAGTCAGATGTGAAATTCCCGGGCTTAACCTGGGCGCTGCATTTGAAACTGGGCATCTAGAGTGCAGGAGAGGAAAGTGGAATTCCTAGTGTAGCGGTGAAATGCGTAGAGATTAGGAAGAACACCAGTGGCGAAGGCGACTTTCTGGACTGTAACTGACACTGAGATACGAAAGCGTGGGTAGCAAACAGGATTAGATACCCTGGTAGTCCACGCCGTAAACGATGAATACTAGGTGTCGGGGGGTACCACCCTCGGTGCCGCAGCAAACGCATTAAGTATTCCGCCTGGGGAGTACGGTCGCAAGATTAAA \n",
      "\n",
      "Marker gene copy: GCA_000175395.3_ASM17539v3_genomic.fna;ACSJ01000001.1;+;351648;1519\n",
      "Forward primer match: ['A[CT]TGGG[CT][AGT]TAAAG[GATC]G', 538]\n",
      "Reverse primer match: ['ACT[CT]AAA[AGT][AG]AATTGACGG', 883]\n",
      "eAmplicon length: 330\n",
      "eAmplicon sequence: TATGTAGGTGGGTGCTTAAGTCAGATGTGAAATTCCCGGGCTTAACCTGGGCGCTGCATTTGAAACTGGGCATCTAGAGTGCAGGAGAGGAAAGTGGAATTCCTAGTGTAGCGGTGAAATGCGTAGAGATTAGGAAGAACACCAGTGGCGAAGGCGACTTTCTGGACTGTAACTGACACTGAGATACGAAAGCGTGGGTAGCAAACAGGATTAGATACCCTGGTAGTCCACGCCGTAAACGATGAATACTAGGTGTCGGGGGGTACCACCCTCGGTGCCGCAGCAAACGCATTAAGTATTCCGCCTGGGGAGTACGGTCGCAAGATTAAA \n",
      "\n",
      "Marker gene copy: GCA_000175395.3_ASM17539v3_genomic.fna;ACSJ01000001.1;+;420998;1519\n",
      "Forward primer match: ['A[CT]TGGG[CT][AGT]TAAAG[GATC]G', 538]\n",
      "Reverse primer match: ['ACT[CT]AAA[AGT][AG]AATTGACGG', 883]\n",
      "eAmplicon length: 330\n",
      "eAmplicon sequence: TATGTAGGTGGGTGCTTAAGTCAGATGTGAAATTCCCGGGCTTAACCTGGGCGCTGCATTTGAAACTGGGCATCTAGAGTGCAGGAGAGGAAAGTGGAATTCCTAGTGTAGCGGTGAAATGCGTAGAGATTAGGAAGAACACCAGTGGCGAAGGCGACTTTCTGGACTGTAACTGACACTGAGATACGAAAGCGTGGGTAGCAAACAGGATTAGATACCCTGGTAGTCCACGCCGTAAACGATGAATACTAGGTGTCGGGGGGTACCACCCTCGGTGCCGCAGCAAACGCATTAAGTATTCCGCCTGGGGAGTACGGTCGCAAGATTAAA \n",
      "\n",
      "Marker gene copy: GCA_000175395.3_ASM17539v3_genomic.fna;ACSJ01000001.1;+;437951;1519\n",
      "Forward primer match: ['A[CT]TGGG[CT][AGT]TAAAG[GATC]G', 538]\n",
      "Reverse primer match: ['ACT[CT]AAA[AGT][AG]AATTGACGG', 883]\n",
      "eAmplicon length: 330\n",
      "eAmplicon sequence: TATGTAGGTGGGTGCTTAAGTCAGATGTGAAATTCCCGGGCTTAACCTGGGCGCTGCATTTGAAACTGGGCATCTAGAGTGCAGGAGAGGAAAGTGGAATTCCTAGTGTAGCGGTGAAATGCGTAGAGATTAGGAAGAACACCAGTGGCGAAGGCGACTTTCTGGACTGTAACTGACACTGAGATACGAAAGCGTGGGTAGCAAACAGGATTAGATACCCTGGTAGTCCACGCCGTAAACGATGAATACTAGGTGTCGGGGGGTACCACCCTCGGTGCCGCAGCAAACGCATTAAGTATTCCGCCTGGGGAGTACGGTCGCAAGATTAAA \n",
      "\n"
     ]
    }
   ],
   "source": [
    "mk_dict = ePCR(f_pr,r_pr,rRNA_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "efficient-grave",
   "metadata": {},
   "source": [
    "### Use eAmplicons and dataset representative sequences fasta to find matches between amplicons and genomes in the collection."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "swiss-vocabulary",
   "metadata": {},
   "outputs": [],
   "source": [
    "mk_seqs = \"/Users/kbeilsmith/Desktop/2024_AmtG/AmtG_0.0.2_test_files/rep_seqs.fasta\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "seventh-three",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Amplicons in dataset: 10 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "count = 0\n",
    "for rep_seq in SeqIO.parse(mk_seqs, \"fasta\"):\n",
    "    count = count+1\n",
    "print(f\"Amplicons in dataset: {count} \\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "accredited-church",
   "metadata": {},
   "source": [
    "##### Set parameters for pairwise alignment\n",
    "\n",
    "mode: 'global'\n",
    "\n",
    "match_score = 1\n",
    "\n",
    "mismatch_score = 0\n",
    "\n",
    "open_gap_score = -1\n",
    "\n",
    "extend_gap_score = -0.5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "billion-south",
   "metadata": {},
   "outputs": [],
   "source": [
    "aligner = Align.PairwiseAligner(mode='global', match_score=1, mismatch_score=0, open_gap_score = -1, extend_gap_score = -0.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "nasty-string",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Given a file of amplicon sequences and a set of eAmplicons generated from from genomes,\n",
    "# this function returns a record of the marker gene copies matching each amplicon,\n",
    "# which includes genome name and location.\n",
    "\n",
    "def fuzzy_match_amplicons_to_genomes(rep_seqs,eAmplicon_set,identity_threshold):\n",
    "    \n",
    "    percent_identical = identity_threshold\n",
    "    # print(f\"Identity threshold {percent_identical}. \\n\")\n",
    "    percent_mismatch = 1 - percent_identical\n",
    "    \n",
    "    AmtG_dict = {}\n",
    "    \n",
    "    # For each amplicon from the dataset:\n",
    "    for amplicon in SeqIO.parse(rep_seqs, \"fasta\"):\n",
    "        amplicon_id = f\"{amplicon.id}\"\n",
    "        print(f\"Amplicon: {amplicon.id}\")\n",
    "        \n",
    "        AmtG_list = []\n",
    "        # For each eAmplicon from the genomes:\n",
    "        for eAmplicon in eAmplicon_set:\n",
    "            \n",
    "            # Check for undefined nucleotides.\n",
    "            if eAmplicon_set[eAmplicon].find(\"N\")< 0:\n",
    "                \n",
    "                # Find the number of expected matching nucleotides for this threshold.\n",
    "                amplicon_length = len(eAmplicon_set[eAmplicon])\n",
    "                expected_matches = percent_identical*amplicon_length\n",
    "                expected_matches_round = round(expected_matches)\n",
    "                # print(f\"Expected matches = {expected_matches_round} of {amplicon_length} nucleotides.\")\n",
    "                \n",
    "                # Generate a scored alignment.\n",
    "                alignment_score = aligner.score(amplicon.seq, eAmplicon_set[eAmplicon])\n",
    "                alignment_score_round = round(alignment_score)\n",
    "                #print(f\"Alignment score = {alignment_score_round}.\")\n",
    "                \n",
    "                # Compare the alignment score and expected matches.\n",
    "                if alignment_score_round >= expected_matches_round:\n",
    "                    # print(f\"MATCH: alignment score {alignment_score_round} exceeds expected score {expected_matches_round}.\")\n",
    "                    # Add matching 16S copies to list.\n",
    "                    AmtG_list.append(eAmplicon)\n",
    "            \n",
    "        match_count = len(AmtG_list)\n",
    "        print(f\"Matches: {match_count}\")           \n",
    "                #else:\n",
    "                    #print(f\"NO MATCH: alignment score {alignment_score_round} below expected score {expected_matches_round}.\")\n",
    "            #else:\n",
    "                #print(f\"Reference amplicon has undefined nucleotides.\")\n",
    "                \n",
    "        # Create entry with amplicon ID and matching genomes list.\n",
    "        AmtG_dict[amplicon_id] = AmtG_list\n",
    "\n",
    "    return AmtG_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "greenhouse-france",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Amplicon: 001cfeeb86f51bd8ae2b79268c8d806d\n",
      "Matches: 0\n",
      "Amplicon: 003902c5913081651d13fa0042ecac85\n",
      "Matches: 0\n",
      "Amplicon: 003ef22a3dcf8c6e655f6fcb9e5b8883\n",
      "Matches: 0\n",
      "Amplicon: 004d919d8ff500a00bfe768062d3e30a\n",
      "Matches: 0\n",
      "Amplicon: 006acab58c360970518d08544e77f20f\n",
      "Matches: 0\n",
      "Amplicon: 0073d25591f05f6cfc230a89f51cd841\n",
      "Matches: 0\n",
      "Amplicon: 007a974b0637bc17763c30e89fbe0fec\n",
      "Matches: 0\n",
      "Amplicon: 008fb713223dbad791ce1c39c7d964b3\n",
      "Matches: 0\n",
      "Amplicon: 00a930c5686229958d943fd2a962c3ef\n",
      "Matches: 0\n",
      "Amplicon: 00bd2dd899758eea2b046d677293459a\n",
      "Matches: 2\n"
     ]
    }
   ],
   "source": [
    "mk_matches = fuzzy_match_amplicons_to_genomes(mk_seqs,mk_dict,.95)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "coated-champion",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'001cfeeb86f51bd8ae2b79268c8d806d': [],\n",
       " '003902c5913081651d13fa0042ecac85': [],\n",
       " '003ef22a3dcf8c6e655f6fcb9e5b8883': [],\n",
       " '004d919d8ff500a00bfe768062d3e30a': [],\n",
       " '006acab58c360970518d08544e77f20f': [],\n",
       " '0073d25591f05f6cfc230a89f51cd841': [],\n",
       " '007a974b0637bc17763c30e89fbe0fec': [],\n",
       " '008fb713223dbad791ce1c39c7d964b3': [],\n",
       " '00a930c5686229958d943fd2a962c3ef': [],\n",
       " '00bd2dd899758eea2b046d677293459a': ['GCF_000157535.1_ASM15753v1_genomic.fna;NZ_GG688487.1;-;2137;1573',\n",
       "  'GCF_000157555.1_ASM15755v1_genomic.fna;NZ_GG688445.1;+;133;1573']}"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mk_matches"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "celtic-background",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Given a file of amplicon sequences and a set of eAmplicons generated from from genomes,\n",
    "# this function returns a record of the marker gene copies matching each amplicon,\n",
    "# which includes genome name and location.\n",
    "\n",
    "def best_match_amplicons_to_genomes(rep_seqs,eAmplicon_set,identity_threshold):\n",
    "    \n",
    "    percent_identical = identity_threshold\n",
    "    print(f\"Finding matches with at least {percent_identical} identity. \\n\")   \n",
    "    identity_range = np.flip(np.arange(percent_identical, 1.01, 0.01).round(2))\n",
    "    \n",
    "    AmtG_best_matches = {}\n",
    "    \n",
    "    for i in identity_range:\n",
    "        print(\" \")\n",
    "        print(f\"Running identity threshold {i}.\")\n",
    "            \n",
    "        matches = fuzzy_match_amplicons_to_genomes(rep_seqs,eAmplicon_set,i)\n",
    "        matched_ASVs = {ASV:hits for (ASV,hits) in matches.items() if len(hits) > 0}\n",
    "        matched_ASV_names = list(matched_ASVs.keys())\n",
    "        #print(f\"Matched ASVs: {matched_ASV_names}.\")\n",
    "    \n",
    "        for (ASV,hits) in matches.items():\n",
    "            # print(ASV)\n",
    "            if ASV not in AmtG_best_matches.keys():\n",
    "                # print(hits)\n",
    "                if len(hits) > 0:\n",
    "                    print(f\"Adding hits for {ASV}.\")\n",
    "                    # print(ASV)\n",
    "                    # print(hits)\n",
    "                    AmtG_best_matches[ASV] = [i,hits]\n",
    "            elif ASV in AmtG_best_matches.keys():\n",
    "                print(f\"{ASV} already has a best hit.\")\n",
    "\n",
    "    return AmtG_best_matches"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "referenced-laptop",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Finding matches with at least 0.9 identity. \n",
      "\n",
      " \n",
      "Running identity threshold 1.0.\n",
      "Amplicon: 001cfeeb86f51bd8ae2b79268c8d806d\n",
      "Matches: 0\n",
      "Amplicon: 003902c5913081651d13fa0042ecac85\n",
      "Matches: 0\n",
      "Amplicon: 003ef22a3dcf8c6e655f6fcb9e5b8883\n",
      "Matches: 0\n",
      "Amplicon: 004d919d8ff500a00bfe768062d3e30a\n",
      "Matches: 0\n",
      "Amplicon: 006acab58c360970518d08544e77f20f\n",
      "Matches: 0\n",
      "Amplicon: 0073d25591f05f6cfc230a89f51cd841\n",
      "Matches: 0\n",
      "Amplicon: 007a974b0637bc17763c30e89fbe0fec\n",
      "Matches: 0\n",
      "Amplicon: 008fb713223dbad791ce1c39c7d964b3\n",
      "Matches: 0\n",
      "Amplicon: 00a930c5686229958d943fd2a962c3ef\n",
      "Matches: 0\n",
      "Amplicon: 00bd2dd899758eea2b046d677293459a\n",
      "Matches: 0\n",
      " \n",
      "Running identity threshold 0.99.\n",
      "Amplicon: 001cfeeb86f51bd8ae2b79268c8d806d\n",
      "Matches: 0\n",
      "Amplicon: 003902c5913081651d13fa0042ecac85\n",
      "Matches: 0\n",
      "Amplicon: 003ef22a3dcf8c6e655f6fcb9e5b8883\n",
      "Matches: 0\n",
      "Amplicon: 004d919d8ff500a00bfe768062d3e30a\n",
      "Matches: 0\n",
      "Amplicon: 006acab58c360970518d08544e77f20f\n",
      "Matches: 0\n",
      "Amplicon: 0073d25591f05f6cfc230a89f51cd841\n",
      "Matches: 0\n",
      "Amplicon: 007a974b0637bc17763c30e89fbe0fec\n",
      "Matches: 0\n",
      "Amplicon: 008fb713223dbad791ce1c39c7d964b3\n",
      "Matches: 0\n",
      "Amplicon: 00a930c5686229958d943fd2a962c3ef\n",
      "Matches: 0\n",
      "Amplicon: 00bd2dd899758eea2b046d677293459a\n",
      "Matches: 0\n",
      " \n",
      "Running identity threshold 0.98.\n",
      "Amplicon: 001cfeeb86f51bd8ae2b79268c8d806d\n",
      "Matches: 0\n",
      "Amplicon: 003902c5913081651d13fa0042ecac85\n",
      "Matches: 0\n",
      "Amplicon: 003ef22a3dcf8c6e655f6fcb9e5b8883\n",
      "Matches: 0\n",
      "Amplicon: 004d919d8ff500a00bfe768062d3e30a\n",
      "Matches: 0\n",
      "Amplicon: 006acab58c360970518d08544e77f20f\n",
      "Matches: 0\n",
      "Amplicon: 0073d25591f05f6cfc230a89f51cd841\n",
      "Matches: 0\n",
      "Amplicon: 007a974b0637bc17763c30e89fbe0fec\n",
      "Matches: 0\n",
      "Amplicon: 008fb713223dbad791ce1c39c7d964b3\n",
      "Matches: 0\n",
      "Amplicon: 00a930c5686229958d943fd2a962c3ef\n",
      "Matches: 0\n",
      "Amplicon: 00bd2dd899758eea2b046d677293459a\n",
      "Matches: 0\n",
      " \n",
      "Running identity threshold 0.97.\n",
      "Amplicon: 001cfeeb86f51bd8ae2b79268c8d806d\n",
      "Matches: 0\n",
      "Amplicon: 003902c5913081651d13fa0042ecac85\n",
      "Matches: 0\n",
      "Amplicon: 003ef22a3dcf8c6e655f6fcb9e5b8883\n",
      "Matches: 0\n",
      "Amplicon: 004d919d8ff500a00bfe768062d3e30a\n",
      "Matches: 0\n",
      "Amplicon: 006acab58c360970518d08544e77f20f\n",
      "Matches: 0\n",
      "Amplicon: 0073d25591f05f6cfc230a89f51cd841\n",
      "Matches: 0\n",
      "Amplicon: 007a974b0637bc17763c30e89fbe0fec\n",
      "Matches: 0\n",
      "Amplicon: 008fb713223dbad791ce1c39c7d964b3\n",
      "Matches: 0\n",
      "Amplicon: 00a930c5686229958d943fd2a962c3ef\n",
      "Matches: 0\n",
      "Amplicon: 00bd2dd899758eea2b046d677293459a\n",
      "Matches: 0\n",
      " \n",
      "Running identity threshold 0.96.\n",
      "Amplicon: 001cfeeb86f51bd8ae2b79268c8d806d\n",
      "Matches: 0\n",
      "Amplicon: 003902c5913081651d13fa0042ecac85\n",
      "Matches: 0\n",
      "Amplicon: 003ef22a3dcf8c6e655f6fcb9e5b8883\n",
      "Matches: 0\n",
      "Amplicon: 004d919d8ff500a00bfe768062d3e30a\n",
      "Matches: 0\n",
      "Amplicon: 006acab58c360970518d08544e77f20f\n",
      "Matches: 0\n",
      "Amplicon: 0073d25591f05f6cfc230a89f51cd841\n",
      "Matches: 0\n",
      "Amplicon: 007a974b0637bc17763c30e89fbe0fec\n",
      "Matches: 0\n",
      "Amplicon: 008fb713223dbad791ce1c39c7d964b3\n",
      "Matches: 0\n",
      "Amplicon: 00a930c5686229958d943fd2a962c3ef\n",
      "Matches: 0\n",
      "Amplicon: 00bd2dd899758eea2b046d677293459a\n",
      "Matches: 0\n",
      " \n",
      "Running identity threshold 0.95.\n",
      "Amplicon: 001cfeeb86f51bd8ae2b79268c8d806d\n",
      "Matches: 0\n",
      "Amplicon: 003902c5913081651d13fa0042ecac85\n",
      "Matches: 0\n",
      "Amplicon: 003ef22a3dcf8c6e655f6fcb9e5b8883\n",
      "Matches: 0\n",
      "Amplicon: 004d919d8ff500a00bfe768062d3e30a\n",
      "Matches: 0\n",
      "Amplicon: 006acab58c360970518d08544e77f20f\n",
      "Matches: 0\n",
      "Amplicon: 0073d25591f05f6cfc230a89f51cd841\n",
      "Matches: 0\n",
      "Amplicon: 007a974b0637bc17763c30e89fbe0fec\n",
      "Matches: 0\n",
      "Amplicon: 008fb713223dbad791ce1c39c7d964b3\n",
      "Matches: 0\n",
      "Amplicon: 00a930c5686229958d943fd2a962c3ef\n",
      "Matches: 0\n",
      "Amplicon: 00bd2dd899758eea2b046d677293459a\n",
      "Matches: 2\n",
      "Adding hits for 00bd2dd899758eea2b046d677293459a.\n",
      " \n",
      "Running identity threshold 0.94.\n",
      "Amplicon: 001cfeeb86f51bd8ae2b79268c8d806d\n",
      "Matches: 0\n",
      "Amplicon: 003902c5913081651d13fa0042ecac85\n",
      "Matches: 0\n",
      "Amplicon: 003ef22a3dcf8c6e655f6fcb9e5b8883\n",
      "Matches: 0\n",
      "Amplicon: 004d919d8ff500a00bfe768062d3e30a\n",
      "Matches: 0\n",
      "Amplicon: 006acab58c360970518d08544e77f20f\n",
      "Matches: 0\n",
      "Amplicon: 0073d25591f05f6cfc230a89f51cd841\n",
      "Matches: 0\n",
      "Amplicon: 007a974b0637bc17763c30e89fbe0fec\n",
      "Matches: 0\n",
      "Amplicon: 008fb713223dbad791ce1c39c7d964b3\n",
      "Matches: 0\n",
      "Amplicon: 00a930c5686229958d943fd2a962c3ef\n",
      "Matches: 0\n",
      "Amplicon: 00bd2dd899758eea2b046d677293459a\n",
      "Matches: 2\n",
      "00bd2dd899758eea2b046d677293459a already has a best hit.\n",
      " \n",
      "Running identity threshold 0.93.\n",
      "Amplicon: 001cfeeb86f51bd8ae2b79268c8d806d\n",
      "Matches: 0\n",
      "Amplicon: 003902c5913081651d13fa0042ecac85\n",
      "Matches: 0\n",
      "Amplicon: 003ef22a3dcf8c6e655f6fcb9e5b8883\n",
      "Matches: 0\n",
      "Amplicon: 004d919d8ff500a00bfe768062d3e30a\n",
      "Matches: 0\n",
      "Amplicon: 006acab58c360970518d08544e77f20f\n",
      "Matches: 0\n",
      "Amplicon: 0073d25591f05f6cfc230a89f51cd841\n",
      "Matches: 0\n",
      "Amplicon: 007a974b0637bc17763c30e89fbe0fec\n",
      "Matches: 0\n",
      "Amplicon: 008fb713223dbad791ce1c39c7d964b3\n",
      "Matches: 0\n",
      "Amplicon: 00a930c5686229958d943fd2a962c3ef\n",
      "Matches: 0\n",
      "Amplicon: 00bd2dd899758eea2b046d677293459a\n",
      "Matches: 2\n",
      "00bd2dd899758eea2b046d677293459a already has a best hit.\n",
      " \n",
      "Running identity threshold 0.92.\n",
      "Amplicon: 001cfeeb86f51bd8ae2b79268c8d806d\n",
      "Matches: 0\n",
      "Amplicon: 003902c5913081651d13fa0042ecac85\n",
      "Matches: 0\n",
      "Amplicon: 003ef22a3dcf8c6e655f6fcb9e5b8883\n",
      "Matches: 0\n",
      "Amplicon: 004d919d8ff500a00bfe768062d3e30a\n",
      "Matches: 0\n",
      "Amplicon: 006acab58c360970518d08544e77f20f\n",
      "Matches: 0\n",
      "Amplicon: 0073d25591f05f6cfc230a89f51cd841\n",
      "Matches: 0\n",
      "Amplicon: 007a974b0637bc17763c30e89fbe0fec\n",
      "Matches: 1\n",
      "Amplicon: 008fb713223dbad791ce1c39c7d964b3\n",
      "Matches: 0\n",
      "Amplicon: 00a930c5686229958d943fd2a962c3ef\n",
      "Matches: 0\n",
      "Amplicon: 00bd2dd899758eea2b046d677293459a\n",
      "Matches: 2\n",
      "Adding hits for 007a974b0637bc17763c30e89fbe0fec.\n",
      "00bd2dd899758eea2b046d677293459a already has a best hit.\n",
      " \n",
      "Running identity threshold 0.91.\n",
      "Amplicon: 001cfeeb86f51bd8ae2b79268c8d806d\n",
      "Matches: 0\n",
      "Amplicon: 003902c5913081651d13fa0042ecac85\n",
      "Matches: 0\n",
      "Amplicon: 003ef22a3dcf8c6e655f6fcb9e5b8883\n",
      "Matches: 0\n",
      "Amplicon: 004d919d8ff500a00bfe768062d3e30a\n",
      "Matches: 0\n",
      "Amplicon: 006acab58c360970518d08544e77f20f\n",
      "Matches: 0\n",
      "Amplicon: 0073d25591f05f6cfc230a89f51cd841\n",
      "Matches: 0\n",
      "Amplicon: 007a974b0637bc17763c30e89fbe0fec\n",
      "Matches: 1\n",
      "Amplicon: 008fb713223dbad791ce1c39c7d964b3\n",
      "Matches: 0\n",
      "Amplicon: 00a930c5686229958d943fd2a962c3ef\n",
      "Matches: 0\n",
      "Amplicon: 00bd2dd899758eea2b046d677293459a\n",
      "Matches: 2\n",
      "007a974b0637bc17763c30e89fbe0fec already has a best hit.\n",
      "00bd2dd899758eea2b046d677293459a already has a best hit.\n",
      " \n",
      "Running identity threshold 0.9.\n",
      "Amplicon: 001cfeeb86f51bd8ae2b79268c8d806d\n",
      "Matches: 0\n",
      "Amplicon: 003902c5913081651d13fa0042ecac85\n",
      "Matches: 0\n",
      "Amplicon: 003ef22a3dcf8c6e655f6fcb9e5b8883\n",
      "Matches: 0\n",
      "Amplicon: 004d919d8ff500a00bfe768062d3e30a\n",
      "Matches: 1\n",
      "Amplicon: 006acab58c360970518d08544e77f20f\n",
      "Matches: 0\n",
      "Amplicon: 0073d25591f05f6cfc230a89f51cd841\n",
      "Matches: 0\n",
      "Amplicon: 007a974b0637bc17763c30e89fbe0fec\n",
      "Matches: 6\n",
      "Amplicon: 008fb713223dbad791ce1c39c7d964b3\n",
      "Matches: 0\n",
      "Amplicon: 00a930c5686229958d943fd2a962c3ef\n",
      "Matches: 0\n",
      "Amplicon: 00bd2dd899758eea2b046d677293459a\n",
      "Matches: 2\n",
      "Adding hits for 004d919d8ff500a00bfe768062d3e30a.\n",
      "007a974b0637bc17763c30e89fbe0fec already has a best hit.\n",
      "00bd2dd899758eea2b046d677293459a already has a best hit.\n"
     ]
    }
   ],
   "source": [
    "best_matches = best_match_amplicons_to_genomes(mk_seqs,mk_dict,.90)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "fabulous-membership",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'00bd2dd899758eea2b046d677293459a': [0.95,\n",
       "  ['GCF_000157535.1_ASM15753v1_genomic.fna;NZ_GG688487.1;-;2137;1573',\n",
       "   'GCF_000157555.1_ASM15755v1_genomic.fna;NZ_GG688445.1;+;133;1573']],\n",
       " '007a974b0637bc17763c30e89fbe0fec': [0.92,\n",
       "  ['GCA_000210035.1_ASM21003v1_genomic.fna;FP929055.1;+;147;1539']],\n",
       " '004d919d8ff500a00bfe768062d3e30a': [0.9,\n",
       "  ['GCA_000210035.1_ASM21003v1_genomic.fna;FP929055.1;+;147;1539']]}"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "best_matches"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "applicable-ghost",
   "metadata": {},
   "source": [
    "### Write out mappings JSON file. This file is used as input for the community modeling pipeline."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "hazardous-reporter",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'00bd2dd899758eea2b046d677293459a': [0.95, ['GCF_000157535.1', 'GCF_000157555.1']], '007a974b0637bc17763c30e89fbe0fec': [0.92, ['GCA_000210035.1']], '004d919d8ff500a00bfe768062d3e30a': [0.9, ['GCA_000210035.1']]}\n"
     ]
    }
   ],
   "source": [
    "# Reformat to store only threshold and unique genome names for the matches.\n",
    "\n",
    "out_best_matches = {}\n",
    "\n",
    "for match in best_matches:\n",
    "    # print(best_matches[match][1])\n",
    "    genome_list = []\n",
    "    for entry in best_matches[match][1]:\n",
    "        #print(entry.split(\"_\")[0:2])\n",
    "        #print(entry.split(\"_\")[0:2])\n",
    "        entry_name = entry.split(\"_\")[0:2]\n",
    "        # print(\"_\".join(entry_name))\n",
    "        entry_name = \"_\".join(entry_name)\n",
    "        genome_list.append(entry_name)\n",
    "    out_best_matches[match] = [best_matches[match][0], list(set(genome_list))]\n",
    "    \n",
    "print(out_best_matches)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "sealed-purple",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write out\n",
    "\n",
    "output_json = \"/Users/kbeilsmith/Desktop/2024_AmtG/AmtG_0.0.2_test_files/test_best_mappings.json\"\n",
    "with open(output_json, \"w\") as outfile:\n",
    "    json.dump(out_best_matches, outfile)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
